flexmeasures/data/models/planning/linear_optimization.py

```diff
@@ -191,8 +210,8 @@ def device_derivative_up_efficiency(m, d, j):
             return 1
         return eff
 
-    model.up_price = Param(model.c, model.j, initialize=price_up_select)
-    model.down_price = Param(model.c, model.j, initialize=price_down_select)
+    model.up_price = Param(model.u, model.c, model.j, initialize=price_up_select)
```

Comment:
Using 'u','p' allows the user to have different number or 'consumption_price_sensor_per device' and 'production_price_sensor_per_device'. Using 'd' would not serve that purpose.

Reply:
Correct me if I'm wrong, the idea of this PR is to allow having different price sensor between **devices**.  That means that we want to map each device to a price sensor.

For example:
```
Device 1  -> Consumption 1, Production 2
Device 2  -> Consumption 3, Production 4
Device 3  -> Consumption 5, Production 6
```

In this example, we would have `consumption_price_sensor_per_device = [1,3,5]` and `production_price_sensor_per_device = [2,4,6]`.

Having the `consumption_price_sensor_per_device[d]' indexed by the device, we could get the corresponding price sensor to each device. This is, of course, assuming that we can have only 2 price sensors per device (consumption and production).

Another interpretation of having  'allow having different price sensor between **devices** could be to allow defining multiple price sensors to a device. In that case, there are different ways to have multiple price sensors "attached" to a device: sum, different times of the day. Is this later interpretation the one that you had in mind?

Reply:
My interpretation was on the basis of the assumption that a device may have only price sensor attached to it.For example,

1. Solar Plants will have only Production Price Sensor as they ideally wont be consuming electricity
2. Load/Building will have consumption price sensor only as they are supposed to only consume electricity.

Let me know if my understanding is right.

Reply:
What do you think? @victorgarcia98

Reply:
Thanks, we are on the same page!

Then, we can use the device index `d` (instead of `u` and `p`) to get which price time series correspond to a particular device, given that there would be at most one.

Given a dictionary that maps devices to the prices, we can iterate through all the devices and in case the price is missing in the dictionary, we just have one with price zero, as @Flix6x mentioned in another conversation.

Reply:
Sure @victorgarcia98 .That works.
But couple of things i needed to clarify before working it through:

1. Is there any particular problem with the introduction of the new index dimensions?
2. For instance,grid is a flexible device.So in that case ,I wont be able to map price sensors to the grid because 'd' here refers to the set of inflexible devices.How will that be taken care of?
3. Will putting price zero to some price sensor mean that the device can consume/produce at cost 0 or will it restrict the consumption/production of energy for that device(in case consumption/production is set to zero)?According to my understanding it will be the first case and that will make our model give different results.

Please let me know if you understood my concerns.Also let me know that you think?

Reply:
2. Applying prices to the power flow exchanged with the grid is exactly the use case we currently support (and want to keep supporting). We do not explicitly model the grid as a separate device, but rather model it implicitly as the sum of (the power flow of) all devices. Grid (power) constraints can be set through the variable `ems_constraints`. Grid prices can be set separately for consumption and production, using `consumption_price_sensor` and `production_price_sensor`, respectively.

3. A zero device consumption price would indeed mean that the device consumes without accruing costs, and a zero device production price would mean that the device produces without accruing revenue. The zero price should not constraint their power flow. But their power flow may still lead to accruing costs or revenues in case there are non-zero prices on the grid.


Hope this clarifies things?

Reply:
Yes thanks @Flix6x.Things are clearer now.
I will change the dependency of the price sensors to 'd' instead then.

Reply:
Hi again @victorgarcia98 @Flix6x 
Upon thinking it through even more I realized that `d` index wont suffice for the situation where i have a building/load as one of my inflexible devices. 
In this case,the prices for consumption are not to be set and are to be decided by the solver.
But according to what @Flix6x said we could either set the prices or set it to 0 while iterating through the inflexible devices.This wont work out here.
I really think there is the necessity for these new index dimensions,which otherwise would just complicate stuff.
We could get on a call if you are still not satisfied with the reasoning.